File: src/models/bkt.ts
import type { ResponseEvent } from '../core/event-schema.js';
import type {
  FittedModel,
  MasteryLearnerEntry,
  MasteryReport,
  ScoringModel,
} from '../core/scoring-model.js';

export interface BktParams {
  /** Prior probability the learner already knows the skill before any evidence. */
  pInit: number;
  /** Probability of learning the skill on any given opportunity (per response). */
  pTransit: number;
  /** Probability of an incorrect response despite knowing the skill. */
  pSlip: number;
  /** Probability of a correct response despite not knowing the skill. */
  pGuess: number;
}

/** Standard textbook default BKT parameters. */
export const BKT_DEFAULT_PARAMS: BktParams = {
  pInit: 0.4,
  pTransit: 0.3,
  pSlip: 0.1,
  pGuess: 0.2,
};

export interface BktConfig {
  /** Overrides applied on top of BKT_DEFAULT_PARAMS for every skill. */
  defaultParams?: Partial<BktParams>;
  /** Per-skill parameter overrides, keyed by skillId. Wins over `fit`. */
  skillParams?: Record<string, Partial<BktParams>>;
  /**
   * When true, any skill without an explicit `skillParams` override has its
   * parameters fit from the observed data via grid search instead of using
   * fixed defaults.
   */
  fit?: boolean;
}

interface BktSkillResult {
  learnerId: string;
  skillId: string;
  /** P(L_t | observation) after each response, in chronological order. */
  posteriorHistory: number[];
  /** Mastery estimate given all evidence observed so far (last posterior). */
  finalMastery: number;
  responseCount: number;
}

export interface BktFittedModel extends FittedModel {
  modelName: 'bkt';
  params: Record<string, BktParams>;
  results: BktSkillResult[];
}

// A control character that will never legitimately appear in a learnerId or
// skillId, used to join/split the composite grouping key safely.
const LEARNER_SKILL_SEP = '\u0000';

function groupByLearnerSkill(events: ResponseEvent[]): Map<string, ResponseEvent[]> {
  const groups = new Map<string, ResponseEvent[]>();
  for (const event of events) {
    const key = `${event.learnerId}${LEARNER_SKILL_SEP}${event.skillId}`;
    const list = groups.get(key);
    if (list) {
      list.push(event);
    } else {
      groups.set(key, [event]);
    }
  }
  for (const list of groups.values()) {
    list.sort((a, b) => Date.parse(a.timestamp) - Date.parse(b.timestamp));
  }
  return groups;
}

function groupBySkill(events: ResponseEvent[]): Map<string, ResponseEvent[]> {
  const groups = new Map<string, ResponseEvent[]>();
  for (const event of events) {
    const list = groups.get(event.skillId);
    if (list) {
      list.push(event);
    } else {
      groups.set(event.skillId, [event]);
    }
  }
  return groups;
}

interface ForwardRecursionDetail {
  /** P(L_t | observation) after each response. */
  posteriorHistory: number[];
  /** Predicted P(correct) at each step, computed from the pre-update prior. */
  predictedProbabilities: number[];
}

/**
 * Runs the BKT forward recursion for one chronologically ordered sequence
 * of responses, given fixed parameters:
 *
 *   P(L_0) = pInit
 *   after correct:   P(L_t|obs) = P(L_t)*(1-pSlip) / [P(L_t)*(1-pSlip) + (1-P(L_t))*pGuess]
 *   after incorrect: P(L_t|obs) = P(L_t)*pSlip     / [P(L_t)*pSlip     + (1-P(L_t))*(1-pGuess)]
 *   P(L_{t+1}) = P(L_t|obs) + (1 - P(L_t|obs)) * pTransit
 *
 * Also records, for each step, the predicted P(correct) computed from the
 * pre-update prior (used by the grid-search fit routine to score candidate
 * parameters against observed outcomes).
 */
function runForwardRecursionDetailed(
  responses: boolean[],
  params: BktParams,
): ForwardRecursionDetail {
  const { pInit, pTransit, pSlip, pGuess } = params;
  let priorL = pInit;
  const posteriorHistory: number[] = [];
  const predictedProbabilities: number[] = [];

  for (const correct of responses) {
    predictedProbabilities.push(priorL * (1 - pSlip) + (1 - priorL) * pGuess);

    let posterior: number;
    if (correct) {
      const numerator = priorL * (1 - pSlip);
      const denominator = numerator + (1 - priorL) * pGuess;
      posterior = denominator === 0 ? priorL : numerator / denominator;
    } else {
      const numerator = priorL * pSlip;
      const denominator = numerator + (1 - priorL) * (1 - pGuess);
      posterior = denominator === 0 ? priorL : numerator / denominator;
    }
    posteriorHistory.push(posterior);
    priorL = posterior + (1 - posterior) * pTransit;
  }

  return { posteriorHistory, predictedProbabilities };
}

/**
 * Runs the BKT forward recursion for one chronologically ordered sequence
 * of responses and returns the posterior mastery probability P(L_t | obs)
 * after each response. Exported directly (in addition to being used inside
 * BktModel) so the recursion math can be unit-tested against a hand-computed
 * worked example.
 */
export function runForwardRecursion(responses: boolean[], params: BktParams): number[] {
  return runForwardRecursionDetailed(responses, params).posteriorHistory;
}

const GRID_PROBABILITY = [0.05, 0.2, 0.35, 0.5, 0.65, 0.8, 0.95];
// pSlip/pGuess are kept below 0.5 by construction: a "skill" parameter above
// that would mean the mechanism is more often wrong than right, which is not
// a meaningful slip/guess rate in practice.
const GRID_LOW_PROBABILITY = [0.02, 0.1, 0.2, 0.3, 0.4];

/**
 * Coarse grid search over (pInit, pTransit, pSlip, pGuess) for one skill's
 * pooled response sequences, minimizing total squared error between the
 * model's pre-update predicted-correct probability and the actual observed
 * outcome at each step. This is intentionally a simple, dependency-free
 * fitting routine (not a full EM/Baum-Welch implementation) that is good
 * enough to noticeably improve on the textbook defaults for a given dataset.
 */
export function fitSkillParamsByGridSearch(sequences: boolean[][]): BktParams {
  let best: BktParams = { ...BKT_DEFAULT_PARAMS };
  let bestError = Infinity;

  for (const pInit of GRID_PROBABILITY) {
    for (const pTransit of GRID_PROBABILITY) {
      for (const pSlip of GRID_LOW_PROBABILITY) {
        for (const pGuess of GRID_LOW_PROBABILITY) {
          const params: BktParams = { pInit, pTransit, pSlip, pGuess };
          let error = 0;
          for (const sequence of sequences) {
            const { predictedProbabilities } = runForwardRecursionDetailed(sequence, params);
            for (let i = 0; i < sequence.length; i += 1) {
              const predicted = predictedProbabilities[i] ?? 0;
              const actual = sequence[i] ? 1 : 0;
              error += (predicted - actual) ** 2;
            }
          }
          if (error < bestError) {
            bestError = error;
            best = params;
          }
        }
      }
    }
  }

  return best;
}

/**
 * Bayesian Knowledge Tracing scoring model. Implements the ScoringModel
 * interface so it is interchangeable with IRT from the engine's point of
 * view.
 */
export class BktModel implements ScoringModel<BktFittedModel> {
  public readonly name = 'bkt';

  constructor(private readonly config: BktConfig = {}) {}

  fit(events: ResponseEvent[]): BktFittedModel {
    const bySkill = groupBySkill(events);
    const byLearnerSkill = groupByLearnerSkill(events);

    const params: Record<string, BktParams> = {};
    for (const skillId of bySkill.keys()) {
      const override = this.config.skillParams?.[skillId];
      if (override) {
        params[skillId] = { ...BKT_DEFAULT_PARAMS, ...this.config.defaultParams, ...override };
      } else if (this.config.fit) {
        const sequences: boolean[][] = [];
        const suffix = `${LEARNER_SKILL_SEP}${skillId}`;
        for (const [key, list] of byLearnerSkill.entries()) {
          if (key.endsWith(suffix)) {
            sequences.push(list.map((e) => e.correct));
          }
        }
        params[skillId] = fitSkillParamsByGridSearch(sequences);
      } else {
        params[skillId] = { ...BKT_DEFAULT_PARAMS, ...this.config.defaultParams };
      }
    }

    const results: BktSkillResult[] = [];
    for (const [key, list] of byLearnerSkill.entries()) {
      const [learnerId, skillId] = key.split(LEARNER_SKILL_SEP) as [string, string];
      const skillParams = params[skillId] ?? BKT_DEFAULT_PARAMS;
      const posteriorHistory = runForwardRecursion(
        list.map((e) => e.correct),
        skillParams,
      );
      results.push({
        learnerId,
        skillId,
        posteriorHistory,
        finalMastery: posteriorHistory[posteriorHistory.length - 1] ?? skillParams.pInit,
        responseCount: list.length,
      });
    }

    return { modelName: 'bkt', params, results };
  }

  score(fittedModel: BktFittedModel): MasteryReport {
    const learnerMap = new Map<string, MasteryLearnerEntry>();

    for (const result of fittedModel.results) {
      let entry = learnerMap.get(result.learnerId);
      if (!entry) {
        entry = { learnerId: result.learnerId, skills: [] };
        learnerMap.set(result.learnerId, entry);
      }
      entry.skills.push({
        skillId: result.skillId,
        metric: 'posterior_mastery_probability',
        value: result.finalMastery,
        responseCount: result.responseCount,
        details: {
          posteriorHistory: result.posteriorHistory,
          params: fittedModel.params[result.skillId],
        },
      });
    }

    return {
      model: 'bkt',
      generatedAt: new Date().toISOString(),
      learners: [...learnerMap.values()],
      meta: { params: fittedModel.params },
    };
  }
}

File: python/src/masterytrace/models/bkt.py
"""
Bayesian Knowledge Tracing (BKT) scoring model. Faithful port of
src/models/bkt.ts -- same forward-recursion formulas, same coarse grid
search fitting routine, same default parameters and grid values. No
numerical library is used: BKT's per-response update is four scalar
arithmetic operations, exactly what the TypeScript original does with
plain numbers, so a numpy/scipy dependency would buy nothing here.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, List, Optional

from ..core.event_schema import ResponseEvent
from ..core.scoring_model import FittedModel, MasteryLearnerEntry, MasteryReport, MasterySkillEntry


@dataclass(frozen=True)
class BktParams:
    """
    pInit: prior probability the learner already knows the skill before
      any evidence.
    pTransit: probability of learning the skill on any given opportunity
      (per response).
    pSlip: probability of an incorrect response despite knowing the skill.
    pGuess: probability of a correct response despite not knowing the
      skill.
    """

    p_init: float
    p_transit: float
    p_slip: float
    p_guess: float


BKT_DEFAULT_PARAMS = BktParams(p_init=0.4, p_transit=0.3, p_slip=0.1, p_guess=0.2)


@dataclass
class BktConfig:
    default_params: Optional[Dict[str, float]] = None
    skill_params: Optional[Dict[str, Dict[str, float]]] = None
    fit: bool = False


@dataclass
class BktSkillResult:
    learner_id: str
    skill_id: str
    posterior_history: List[float]
    final_mastery: float
    response_count: int


@dataclass
class BktFittedModel(FittedModel):
    params: Dict[str, BktParams] = field(default_factory=dict)
    results: List[BktSkillResult] = field(default_factory=list)


def _group_by_learner_skill(
    events: List[ResponseEvent],
) -> Dict[str, Dict[str, List[ResponseEvent]]]:
    """
    Groups events by learnerId, then by skillId, sorting each learner+
    skill's events into chronological order. Nested dicts (rather than a
    joined string key) are used deliberately: learnerId and skillId are
    arbitrary user-supplied strings with no character restrictions, so any
    chosen separator could in principle also appear inside an id and
    corrupt the grouping. Nesting sidesteps that entirely -- same
    rationale, same fix, as the TypeScript original.
    """
    groups: Dict[str, Dict[str, List[ResponseEvent]]] = {}
    for event in events:
        by_skill = groups.setdefault(event.learner_id, {})
        by_skill.setdefault(event.skill_id, []).append(event)

    for by_skill in groups.values():
        for skill_id, event_list in by_skill.items():
            by_skill[skill_id] = sorted(event_list, key=lambda e: _parse_timestamp(e.timestamp))
    return groups


def _parse_timestamp(value: str) -> datetime:
    candidate = value[:-1] + "+00:00" if value.endswith("Z") else value
    return datetime.fromisoformat(candidate)


def _group_by_skill(events: List[ResponseEvent]) -> Dict[str, List[ResponseEvent]]:
    groups: Dict[str, List[ResponseEvent]] = {}
    for event in events:
        groups.setdefault(event.skill_id, []).append(event)
    return groups


@dataclass
class _ForwardRecursionDetail:
    posterior_history: List[float]
    predicted_probabilities: List[float]


def _run_forward_recursion_detailed(responses: List[bool], params: BktParams) -> _ForwardRecursionDetail:
    """
    Runs the BKT forward recursion for one chronologically ordered
    sequence of responses, given fixed parameters:

      P(L_0) = pInit
      after correct:   P(L_t|obs) = P(L_t)*(1-pSlip) / [P(L_t)*(1-pSlip) + (1-P(L_t))*pGuess]
      after incorrect: P(L_t|obs) = P(L_t)*pSlip     / [P(L_t)*pSlip     + (1-P(L_t))*(1-pGuess)]
      P(L_{t+1}) = P(L_t|obs) + (1 - P(L_t|obs)) * pTransit
    """
    prior_l = params.p_init
    posterior_history: List[float] = []
    predicted_probabilities: List[float] = []

    for correct in responses:
        predicted_probabilities.append(prior_l * (1 - params.p_slip) + (1 - prior_l) * params.p_guess)

        if correct:
            numerator = prior_l * (1 - params.p_slip)
            denominator = numerator + (1 - prior_l) * params.p_guess
            posterior = prior_l if denominator == 0 else numerator / denominator
        else:
            numerator = prior_l * params.p_slip
            denominator = numerator + (1 - prior_l) * (1 - params.p_guess)
            posterior = prior_l if denominator == 0 else numerator / denominator

        posterior_history.append(posterior)
        prior_l = posterior + (1 - posterior) * params.p_transit

    return _ForwardRecursionDetail(posterior_history=posterior_history, predicted_probabilities=predicted_probabilities)


def run_forward_recursion(responses: List[bool], params: BktParams) -> List[float]:
    """
    Runs the BKT forward recursion for one chronologically ordered
    sequence of responses and returns the posterior mastery probability
    P(L_t | obs) after each response. Exported directly (in addition to
    being used inside BktModel) so the recursion math can be
    unit-tested against a hand-computed worked example.
    """
    return _run_forward_recursion_detailed(responses, params).posterior_history


_GRID_PROBABILITY = [0.05, 0.2, 0.35, 0.5, 0.65, 0.8, 0.95]
# pSlip/pGuess are kept below 0.5 by construction: a "skill" parameter
# above that would mean the mechanism is more often wrong than right,
# which is not a meaningful slip/guess rate in practice.
_GRID_LOW_PROBABILITY = [0.02, 0.1, 0.2, 0.3, 0.4]


def fit_skill_params_by_grid_search(sequences: List[List[bool]]) -> BktParams:
    """
    Coarse grid search over (pInit, pTransit, pSlip, pGuess) for one
    skill's pooled response sequences, minimizing total squared error
    between the model's pre-update predicted-correct probability and the
    actual observed outcome at each step. This is intentionally a simple,
    dependency-free fitting routine (not a full EM/Baum-Welch
    implementation) that is good enough to noticeably improve on the
    textbook defaults for a given dataset -- same trade-off the
    TypeScript original documents and makes.
    """
    best = BKT_DEFAULT_PARAMS
    best_error = float("inf")

    for p_init in _GRID_PROBABILITY:
        for p_transit in _GRID_PROBABILITY:
            for p_slip in _GRID_LOW_PROBABILITY:
                for p_guess in _GRID_LOW_PROBABILITY:
                    params = BktParams(p_init=p_init, p_transit=p_transit, p_slip=p_slip, p_guess=p_guess)
                    error = 0.0
                    for sequence in sequences:
                        detail = _run_forward_recursion_detailed(sequence, params)
                        for i, correct in enumerate(sequence):
                            predicted = detail.predicted_probabilities[i] if i < len(detail.predicted_probabilities) else 0.0
                            actual = 1.0 if correct else 0.0
                            error += (predicted - actual) ** 2
                    if error < best_error:
                        best_error = error
                        best = params

    return best


class BktModel:
    """
    Bayesian Knowledge Tracing scoring model. Implements the ScoringModel
    protocol so it is interchangeable with IRT from the engine's point of
    view.
    """

    name = "bkt"

    def __init__(self, config: Optional[BktConfig] = None) -> None:
        self.config = config or BktConfig()

    def fit(self, events: List[ResponseEvent]) -> BktFittedModel:
        by_skill = _group_by_skill(events)
        by_learner_skill = _group_by_learner_skill(events)

        params: Dict[str, BktParams] = {}
        for skill_id in by_skill:
            override = (self.config.skill_params or {}).get(skill_id)
            if override:
                merged = {
                    "p_init": BKT_DEFAULT_PARAMS.p_init,
                    "p_transit": BKT_DEFAULT_PARAMS.p_transit,
                    "p_slip": BKT_DEFAULT_PARAMS.p_slip,
                    "p_guess": BKT_DEFAULT_PARAMS.p_guess,
                }
                merged.update(self.config.default_params or {})
                merged.update(override)
                params[skill_id] = BktParams(**merged)
            elif self.config.fit:
                sequences: List[List[bool]] = []
                for by_skill_for_learner in by_learner_skill.values():
                    event_list = by_skill_for_learner.get(skill_id)
                    if event_list:
                        sequences.append([e.correct for e in event_list])
                params[skill_id] = fit_skill_params_by_grid_search(sequences)
            else:
                merged = {
                    "p_init": BKT_DEFAULT_PARAMS.p_init,
                    "p_transit": BKT_DEFAULT_PARAMS.p_transit,
                    "p_slip": BKT_DEFAULT_PARAMS.p_slip,
                    "p_guess": BKT_DEFAULT_PARAMS.p_guess,
                }
                merged.update(self.config.default_params or {})
                params[skill_id] = BktParams(**merged)

        results: List[BktSkillResult] = []
        for learner_id, by_skill_for_learner in by_learner_skill.items():
            for skill_id, event_list in by_skill_for_learner.items():
                skill_params = params.get(skill_id, BKT_DEFAULT_PARAMS)
                posterior_history = run_forward_recursion([e.correct for e in event_list], skill_params)
                results.append(
                    BktSkillResult(
                        learner_id=learner_id,
                        skill_id=skill_id,
                        posterior_history=posterior_history,
                        final_mastery=posterior_history[-1] if posterior_history else skill_params.p_init,
                        response_count=len(event_list),
                    )
                )

        return BktFittedModel(model_name="bkt", params=params, results=results)

    def score(self, fitted_model: BktFittedModel) -> MasteryReport:
        learner_map: Dict[str, MasteryLearnerEntry] = {}

        for result in fitted_model.results:
            entry = learner_map.get(result.learner_id)
            if entry is None:
                entry = MasteryLearnerEntry(learner_id=result.learner_id, skills=[])
                learner_map[result.learner_id] = entry
            entry.skills.append(
                MasterySkillEntry(
                    skill_id=result.skill_id,
                    metric="posterior_mastery_probability",
                    value=result.final_mastery,
                    response_count=result.response_count,
                    details={
                        "posterior_history": result.posterior_history,
                        "params": fitted_model.params.get(result.skill_id),
                    },
                )
            )

        return MasteryReport(
            model="bkt",
            generated_at=datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            learners=list(learner_map.values()),
            meta={"params": fitted_model.params},
        )
